### ISSUE TITLE
Arch Linux Bypass the Local File System Cache

### ISSUE THREAD

**Issue by cemalgnlts** [copy/v86#857]
Hi,

**Image:** Arch Linux

I read and run the file written with Node, but after reading the file once, it always gives the same content.

After some research, I found this part in the node document ([File system flags](https://nodejs.org/api/fs.html#file-system-flags)):

> `rs+`: Open file for reading and writing in synchronous mode. Instructs the operating system to bypass the local file system cache.

> This is primarily useful for opening files on NFS mounts as it allows skipping the potentially stale local cache. It has a very real impact on I/O performance so using this flag is not recommended unless it is needed.

Even though I tried this, it still gives the content of the first file I read (it caches the file the first time I read it and always returns this cached file content).

I'm writing the sample file and it's the same result even though I changed its contents. I can understand this is the first time I read the file with `read-start` and `read-end`. Then `read-start` and `read-end` are not triggered.

![Screenshot 2023-05-12 09 58 30](https://github.com/copy/v86/assets/45357531/1305530f-eb49-4e91-9a52-05fead3ba7f9)

I'm not sure if the problem has anything to do with fs9p or anything in the emulator.

**Comment by cemalgnlts** [comment]
I researched the problem a little more. It's not a cache issue. If the emulator.write_file property changes a file a second time, the node does not see it.

![Screenshot 2023-05-12 16 09 38](https://github.com/copy/v86/assets/45357531/65e2c06a-f366-476a-b17e-61926a8a3b66)
![Screenshot 2023-05-12 16 10 15](https://github.com/copy/v86/assets/45357531/a0a7487d-5ecc-42d3-b3b2-5915a946561b)
![Screenshot 2023-05-12 16 10 38](https://github.com/copy/v86/assets/45357531/db8f5ce8-ea74-48a6-9962-ca9b4d9540c0)


I'll do some more research and let you know.

**Comment by cemalgnlts** [comment]
Yes, I was able to test some more;

At first I create a file:
```js
const encoder = new TextEncoder();

emulator.create_file( "/root/index.js", encoder.encode('console.log("test")') );
```

Then I read the file with node, the content comes without any problem.

I then change the content:
```js
const ptr = emulator.fs9p.SearchPath("/root/index.js");
const inode = emulator.fs9p.GetInode(ptr.id);
const buffer = encoder.encode('console.log("Changed!")');

inode.size = buffer.length;

emulator.fs9p.set_data(ptr.id, buffer);
```

If I read the file again with `emulator.read_file` it shows as changed.

But if I try to access the file with node, the contents of the file seem unchanged. If I replace the contents of the file with node: `fs.writeFileSync("index.js", 'console.log("Changed!")')`. Now both node and `fs9p` know that the file has changed.

In summary: If the file is changed with `fs9p` it is not reflected in the emulator.

**Comment by copy** [comment]
Did you try playing around with the `cache` parameter on the kernel command line (https://github.com/copy/v86/blob/0615be5/examples/arch.html#L29)? See https://www.kernel.org/doc/Documentation/filesystems/9p.txt. In particular, the `cache=loose` that is present in the example means that Linux assumes the 9p filesystem doesn't change and doesn't make any effort at writing back to it unless necessary. You may want `cache=mmap` or `cache=none`.
